File: src/main/kotlin/dev/jefy/connectpro/portfolio/applicaion/command/ServiceCommandImpl.java
package dev.jefy.connectpro.portfolio.applicaion.command;


import org.jspecify.annotations.NullMarked;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

import dev.jefy.connectpro.management.ManagementClient;
import dev.jefy.connectpro.management.domain.vo.AwardId;
import dev.jefy.connectpro.management.domain.vo.CategoryId;
import dev.jefy.connectpro.portfolio.applicaion.dtos.FAQRequest;
import dev.jefy.connectpro.portfolio.applicaion.dtos.ServiceRequest;
import dev.jefy.connectpro.portfolio.domain.model.PService;
import dev.jefy.connectpro.portfolio.domain.repository.PortfolioRepository;
import dev.jefy.connectpro.portfolio.domain.repository.ServiceRepository;
import dev.jefy.connectpro.portfolio.domain.vo.FAQId;
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId;
import dev.jefy.connectpro.shared.application.exceptions.ResourceNotFound;
import dev.jefy.connectpro.shared.domain.vo.ImageUrl;
import dev.jefy.connectpro.shared.infrastructure.file_storage.ImageService;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@NullMarked
@Service
@RequiredArgsConstructor
@Transactional
public class ServiceCommandImpl implements ServiceCommand {
    
    private final PortfolioRepository  portfolioRepo;
    private final ServiceRepository serviceRepo;
    private final ImageService imageService;
    private final ManagementClient managementClient;
    
    @Override
    public ServiceId create(ServiceRequest request) {
        PortfolioId portfolioId = PortfolioId.of(request.portfolioId());


        if (managementClient.notExistsCategory(CategoryId.of(request.categoryId()))) {
            throw  new ResourceNotFound("Category " + request.categoryId() + " not found");
        }
        
        if (!portfolioRepo.existsById(portfolioId)) {
            throw  new ResourceNotFound("Portfolio with id " + portfolioId + " not found");
        }
        boolean isConflict = serviceRepo.isTitleConflict(portfolioId, request.title());
        if (isConflict) {
            throw new IllegalStateException("Service title already exists");
        }
        PService service = new PService(portfolioId, request);
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId update(ServiceId serviceId, ServiceRequest request) {
        PService service = getService(serviceId);
        
        if (managementClient.notExistsCategory(CategoryId.of(request.categoryId()))) {
            throw  new ResourceNotFound("Category " + request.categoryId() + " not found");
        }
        
        boolean isConflict = serviceRepo.isTitleConflict(service.getPortfolioId(), request.title());
        if (isConflict) {
            throw new IllegalStateException("Service title already exists");
        }
        service.update(request);
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId setCoverImage(ServiceId serviceId, MultipartFile image) throws IOException {
        Assert.notNull(image, "Image must not be null");
        PService service = getService(serviceId);

        ImageUrl imageUrl = imageService.save(image);
        service.setCoverImageUrl(imageUrl);
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId deleteCoverImage(ServiceId serviceId) throws IOException {
        PService service = getService(serviceId);
        imageService.delete(service.getCoverImageUrl());
        service.deleteCoverImageUrl();
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId addImage(ServiceId serviceId, MultipartFile image) throws IOException {
        Assert.notNull(image, "Image must not be null");
        PService service = getService(serviceId);

        ImageUrl imageUrl = imageService.save(image);
        service.addImage(imageUrl);
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId removeImage(ServiceId serviceId, ImageUrl imageUrl) throws IOException {
        Assert.notNull(imageUrl, "Image must not be null");
        PService service = getService(serviceId);
        imageService.delete(imageUrl);
        service.removeImage(imageUrl);
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId addFaq(ServiceId serviceId, FAQRequest request) {
        Assert.notNull(request, "Faq must not be null");
        PService service = getService(serviceId);
        service.addFaq(request);
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId removeFaq(ServiceId serviceId, FAQId faqId) {
        Assert.notNull(faqId, "Faq Id must not be null");
        PService service = getService(serviceId);
        service.removeFaq(faqId);
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId setAward(ServiceId serviceId, AwardId awardId) {
        Assert.notNull(awardId, "award Id must not be null");


        if (managementClient.notExistsAward(awardId)) {
            throw  new ResourceNotFound("Award " + awardId.value() + " not found");
        }
        
        PService service = getService(serviceId);
        service.setAwardId(awardId);
        serviceRepo.save(service);
        return service.getId();
    }    
    
    @Override
    public ServiceId removeAward(ServiceId serviceId) {
        PService service = getService(serviceId);
        service.removeAward();
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId activateService(ServiceId serviceId) {
        PService service = getService(serviceId);
        service.activate();
        serviceRepo.save(service);
        return service.getId();
    }

    @Override
    public ServiceId deactivateService(ServiceId serviceId) {
        PService service = getService(serviceId);
        service.deactivate();
        serviceRepo.save(service);
        return service.getId();
    }
    
    private PService getService(ServiceId serviceId) {
        Assert.notNull(serviceId, "Service id cannot be null");
        return serviceRepo
                .findById(serviceId)
                .orElseThrow(() -> new ResourceNotFound("Service not found"));
    }
}

File: src/main/kotlin/dev/jefy/connectpro/portfolio/application/command/ServiceCommandImpl.kt
package dev.jefy.connectpro.portfolio.application.command

import dev.jefy.connectpro.engagement.EngagementClient
import dev.jefy.connectpro.management.ManagementClient
import dev.jefy.connectpro.management.domain.vo.AwardId
import dev.jefy.connectpro.management.domain.vo.CategoryId
import dev.jefy.connectpro.portfolio.application.dtos.FAQRequest
import dev.jefy.connectpro.portfolio.application.dtos.ServiceRequest
import dev.jefy.connectpro.portfolio.application.exceptions.AwardNotFoundException
import dev.jefy.connectpro.portfolio.application.exceptions.CategoryNotFoundException
import dev.jefy.connectpro.portfolio.application.exceptions.ServiceAlreadyExistsException
import dev.jefy.connectpro.portfolio.application.exceptions.ServiceNotFoundException
import dev.jefy.connectpro.portfolio.domain.model.Service
import dev.jefy.connectpro.portfolio.domain.repository.PortfolioRepository
import dev.jefy.connectpro.portfolio.domain.repository.ServiceRepository
import dev.jefy.connectpro.portfolio.domain.vo.FAQId
import dev.jefy.connectpro.portfolio.domain.vo.PortfolioId
import dev.jefy.connectpro.portfolio.domain.vo.ServiceId
import dev.jefy.connectpro.shared.application.dtos.ImageData
import dev.jefy.connectpro.shared.domain.vo.Image
import dev.jefy.connectpro.shared.infrastructure.annotations.CommandService
import dev.jefy.connectpro.shared.infrastructure.file_storage.ImageService
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

@CommandService
class ServiceCommandImpl(
    private val portfolioRepo: PortfolioRepository,
    private val serviceRepo: ServiceRepository,
    private val imageService: ImageService,
    private val managementClient: ManagementClient,
    private val engagementClient: EngagementClient,
    private val tagService: TagService
) : ServiceCommand {

    override fun create(request: ServiceRequest): ServiceId {
        val portfolioId = PortfolioId.of(request.portfolioId)
        val categoryId = CategoryId.of(request.categoryId)

        if (managementClient.notExistsCategory(categoryId)) throw CategoryNotFoundException()

        require(portfolioRepo.existsById(portfolioId)) { "Portfolio with id $portfolioId not found" }

        if (serviceRepo.existsByTitleConflict(portfolioId = portfolioId, title = request.name)) throw ServiceAlreadyExistsException()

        val tags = tagService.resolveTags(stringTags = request.tags)

        val service = Service(
            portfolioId = portfolioId,
            name = request.name,
            description = request.description,
            categoryId = categoryId,
            tags = tags.toMutableSet()
        )
        serviceRepo.save(service)

        return service.id
    }

    override fun update(
        serviceId: ServiceId, request: ServiceRequest
    ): ServiceId {
        val service = getService(serviceId)
        val categoryId = CategoryId.of(request.categoryId)

        if (managementClient.notExistsCategory(categoryId)) throw CategoryNotFoundException()

        if (serviceRepo.existsByTitleConflictForId(
                portfolioId = service.portfolioId,
                title = request.name,
                serviceId = serviceId
            )) throw ServiceAlreadyExistsException()

        val tags = tagService.resolveTags(stringTags = request.tags)

        service.update(
            title = request.name,
            description = request.description,
            categoryId = categoryId,
            tags = tags
        )
        serviceRepo.save(service)
        
        return service.id
    }

    @Throws(IOException::class)
    override fun setCoverImage(serviceId: ServiceId, image: MultipartFile): ServiceId {
        val service = getService(serviceId)
        val imageUrl = imageService.save(image)
        service.addCoverImage(imageUrl)
        serviceRepo.save(service)
        return service.id
    }
        

    @Throws(IOException::class)
    override fun deleteCoverImage(serviceId: ServiceId): ServiceId {
        val service = getService(serviceId)
        service.coverImage?.let { imageService.delete(Image(it)) }
        service.deleteCoverImage()
        serviceRepo.save(service)
        return service.id
    }

    @Throws(IOException::class)
    override fun addImage(serviceId: ServiceId, image: MultipartFile): ServiceId {
        val service = getService(serviceId)
        val imageUrl = imageService.save(image)
        service.addImage(imageUrl)
        serviceRepo.save(service)
        return service.id
    }


    @Throws(IOException::class)
    override fun addImages(serviceId: ServiceId, images: List<MultipartFile>): ServiceId {
        val service = getService(serviceId)
        images.forEach { file ->
            val imageUrl = imageService.save(file)
            service.addImage(imageUrl)
        }
        serviceRepo.save(service)
        return service.id
    }

    @Throws(IOException::class)
    override fun removeImage(serviceId: ServiceId, image: ImageData): ServiceId {
        val service = getService(serviceId)
        imageService.delete(image.getKey())
        service.removeImage(image.getKey())
        serviceRepo.save(service)
        return service.id
    }
        

    override fun addFaq(serviceId: ServiceId, request: FAQRequest): ServiceId =
        getService(serviceId)
            .apply { addFaq(request) }
            .also { serviceRepo.save(it) }
            .id

    override fun removeFaq(serviceId: ServiceId, faqId: FAQId): ServiceId =
        getService(serviceId)
            .apply { removeFaq(faqId) }
            .also { serviceRepo.save(it) }
            .id

    override fun setAward(serviceId: ServiceId, awardId: AwardId): ServiceId {
        if (managementClient.notExistsAward(awardId)) throw AwardNotFoundException()

        return getService(serviceId)
            .apply { addAwardId(awardId) }
            .also { serviceRepo.save(it) }
            .id
    }

    override fun removeAward(serviceId: ServiceId): ServiceId =
        getService(serviceId)
            .apply { removeAward() }
            .also { serviceRepo.save(it) }
            .id

    override fun activateService(serviceId: ServiceId): ServiceId =
        getService(serviceId)
            .apply { activate() }
            .also { serviceRepo.save(it) }
            .id

    override fun deactivateService(serviceId: ServiceId): ServiceId =
        getService(serviceId)
            .apply { deactivate() }
            .also { serviceRepo.save(it) }
            .id

    override fun delete(serviceId: ServiceId) {
        val service = getService(serviceId)
        engagementClient.deleteLikesAndReviewsForService(serviceId)
        serviceRepo.deleteById(service.id)
    }

    private fun getService(serviceId: ServiceId): Service = serviceRepo
        .findById(serviceId).orElseThrow { ServiceNotFoundException() }
}